signac/core/synced_collection.py

```diff
@@ -8,19 +8,132 @@
+_BUFFERED_MODE_FORCE_WRITE = None
+_BUFFERED_BACKNDS: List[Any] = list()
+
+
+class BufferException(Error):
+    """An exception occured in buffered mode."""
+
+
+class BufferedError(BufferException):
+    """Raised when an error occured while flushing one or more buffered files.
+
+    Attribute
+    ---------
+    names:
+        A dictionary of names that caused issues during the flush operation,
+        mapped to a possible reason for the issue or None in case that it
+        cannot be determined.
+    """
+
+    def __init__(self, files):
+        self.files = files
+
+    def __str__(self):
+        return "{}({})".format(type(self).__name__, self.files)
+
+
+def flush_all():
+    """Execute all deferred write operations.
+
+    Raises
+    ------
+    BufferedFileError
+    """
+    logger.debug("Flushing buffer...")
+    issues = dict()
+    for backend in _BUFFERED_BACKNDS:
+        try:
+            # try to sync the data to backend
+            issue = backend._flush_buffer()
+            issues.update(issue)
+        except OSError as error:
+            logger.error(str(error))
+            issues[backend] = error
+    if issues:
+        raise BufferedError(issues)
+
+
+def _get_buffer_force_mode():
+    """Return True if buffer force mode enabled."""
+    return _BUFFERED_MODE_FORCE_WRITE
+
+
+def _in_buffered_mode():
+    """Return True if in buffered read/write mode."""
+    return _BUFFERED_MODE > 0
+
+
+def _register_buffered_backend(backend):
+    """Register the backend.
+
+    The registry is used in the :meth:`flush_all`.
+    Every backend to register should have ``_flush_buffer`` method.
+    """
+    _BUFFERED_BACKNDS.append(backend)
+
+
+@contextmanager
+def buffer_reads_writes(force_write=False):
+    """Enter a global buffer mode for all SyncedCollection instances.
+
+    All future write operations are written to the buffer, read
+    operations are performed from the buffer whenever possible.
+
+    All write operations are deferred until the flush_all() function
+    is called, the buffer overflows, or upon exiting the buffer mode.
+
+    Parameters
+    ----------
+    force_write: bool
+        If true, overwrites the metadata check.
+
+    Raises
+    ------
+    BufferException
+    """
+    global _BUFFERED_MODE
+    global _BUFFERED_MODE_FORCE_WRITE
+    assert _BUFFERED_MODE >= 0
+
+    # Can't switch force modes.
+    if _BUFFERED_MODE_FORCE_WRITE is not None and (force_write != _BUFFERED_MODE_FORCE_WRITE):
+        raise BufferException(
+            "Unable to enter buffered mode with force write enabled, because "
+            "we are already in buffered mode with force write disabled and vise-versa.")
+
+    _BUFFERED_MODE_FORCE_WRITE = force_write
+    _BUFFERED_MODE += 1
+    try:
+        yield
+    finally:
+        _BUFFERED_MODE -= 1
+        if _BUFFERED_MODE == 0:
+            try:
+                flush_all()
+            finally:
+                _BUFFERED_MODE_FORCE_WRITE = None
```

Comment:
Should "force write" be cleared ...
1. ... after the context manager in which it was enabled is no longer active, or
2. ... after all buffering is ended?

I believe this code is doing the second (force write persists until all buffering is done) but the first would make more sense to me.

```
enter buffer 0
enter buffer 1 with force write enabled
exit buffer 1
# force write is cleared here in option 1
exit buffer 0
# force write is cleared here in option 2
```

Reply:
I'm going to remove the force write mode entirely with the new implementation. I don't think there's much reason to support it; if we're buffering with a class that implements integrity checks, failure should always be an error. If we see a use case for this in the future, we can reimplement it.

Reply:
There was consensus on Slack, this is fine to remove. Resolving this conversation.